xdsl/transforms/canonicalization_patterns/cf.py

```diff
@@ -172,3 +177,276 @@ def match_and_rewrite(self, op: cf.ConditionalBranch, rewriter: PatternRewriter)
                 op.cond, new_then, new_then_args, new_else, new_else_args
             )
         )
+
+
+class SimplifySwitchWithOnlyDefault(RewritePattern):
+    """
+    switch %flag : i32, [
+      default:  ^bb1
+    ]
+     -> br ^bb1
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: cf.Switch, rewriter: PatternRewriter):
+        if len(op.case_blocks) == 0:
+            rewriter.replace_matched_op(
+                cf.Branch(op.default_block, *op.default_operands)
+            )
+
+
+def drop_case_helper(
+    rewriter: PatternRewriter,
+    op: cf.Switch,
+    predicate: Callable[[AnyIntegerAttr, Block, Sequence[Operation | SSAValue]], bool],
+):
+    case_values = op.case_values
+    if case_values is None:
+        return
+    requires_change = False
+
+    new_case_values: list[int] = []
+    new_case_blocks: list[Block] = []
+    new_case_operands: list[Sequence[Operation | SSAValue]] = []
+
+    for switch_case, block, operands in zip(
+        case_values.data.data, op.case_blocks, op.case_operand
+    ):
+        int_switch_case = cast(AnyIntegerAttr, switch_case)
+        if predicate(int_switch_case, block, operands):
+            requires_change = True
+            continue
+        new_case_values.append(cast(AnyIntegerAttr, switch_case).value.data)
+        new_case_blocks.append(block)
+        new_case_operands.append(operands)
+
+    if requires_change:
+        rewriter.replace_matched_op(
+            cf.Switch(
+                op.flag,
+                op.default_block,
+                op.default_operands,
+                DenseIntOrFPElementsAttr.vector_from_list(
+                    new_case_values, case_values.get_element_type()
+                ),
+                new_case_blocks,
+                new_case_operands,
+            )
+        )
+
+
+class DropSwitchCasesThatMatchDefault(RewritePattern):
+    """
+    switch %flag : i32, [
+      default: ^bb1,
+      42: ^bb1,
+      43: ^bb2
+    ]
+    ->
+    switch %flag : i32, [
+      default: ^bb1,
+      43: ^bb2
+    ]
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: cf.Switch, rewriter: PatternRewriter):
+        def predicate(
+            switch_case: AnyIntegerAttr,
+            block: Block,
+            operands: Sequence[Operation | SSAValue],
+        ) -> bool:
+            return block == op.default_block and operands == op.default_operands
+
+        drop_case_helper(rewriter, op, predicate)
+
+
+def foldSwitch(switch: cf.Switch, rewriter: PatternRewriter, flag: int):
+    """
+    Helper for folding a switch with a constant value.
+    switch %c_42 : i32, [
+      default: ^bb1 ,
+      42: ^bb2,
+      43: ^bb3
+    ]
+    -> br ^bb2
+    """
+    case_values = switch.case_values
+    if case_values is not None:
```

Comment:
does it make sense for this to be optional? Shouldn't it be `()` by default?

Reply:
It is an optional property in mlir. I don't personally think it makes any sense but it is what it is